vault/expiration.go

```diff
@@ -562,11 +611,56 @@ func (m *ExpirationManager) Renew(leaseID string, increment time.Duration) (*log
 	return resp, nil
 }
 
+// RestoreTokenCheck verifies that the token is not expired while running in
+// restore mode.  If we are not in restore mode, the lease has already been
+// restored or the lease still has time left, it returns true.
+func (m *ExpirationManager) RestoreTokenCheck(source string, token string) (bool, error) {
+	defer metrics.MeasureSince([]string{"expire", "restore-token-check"}, time.Now())
+
+	// Return immediately if we are not in retore mode, expiration manager is
+	// already loaded
+	if !m.inRestoreMode() {
+		return true, nil
+	}
+
+	m.restoreModeLock.RLock()
+	defer m.restoreModeLock.RUnlock()
+
+	// Check again after we obtain the lock
+	if !m.inRestoreMode() {
+		return true, nil
+	}
+
+	// Compute the Lease ID
+	saltedID, err := m.tokenStore.SaltID(token)
+	if err != nil {
+		return false, err
+	}
+	leaseID := path.Join(source, saltedID)
+
+	m.lockLease(leaseID)
+	defer m.unlockLease(leaseID)
+
```

Comment:
Now that we have the lease lock I think this needs a check at this point for whether it's in in the restoreLoaded map.

Reply:
Actually, it might belong in loadEntryInternal. Still trying to sort it out.

Reply:
Yeah, I think it belongs in loadEntryInternal IFF we're in restore mode; that way loadEntry still works normally. So, before the map store and the pending update, verify that it's not actually already been inserted into the map and bail if so.

Reply:
I think you are right, we should add the check here.  They only place we care about checking if it is already processed is in the background restore process.   I thought about this early on and decided I didn't care if it was a little slower while restoring and restored the lease multiple times.  Things have evolved since then and it will be an easy enough check to see if that entry has already been processed.

Reply:
As a follow-on I think it means you could get rid of the other check above since it happens just before calling loadEntryInternal.

Reply:
This check has been moved into `loadEntryInternal`.